### ISSUE TITLE
Provides a false sense of security

### ISSUE THREAD

**Issue by acicali** [samayo/bulletproof#90]
Calling this project "BULLETPROOF" and saying "with security" implies to developers that this library will handle all their image uploading security concerns. To an inexperienced developer, this is very dangerous.

This library does not provide much in terms of security for image uploading. For instance, a major security concern of managing image uploads with PHP is that the images may contain malicious code (embedded PHP).

The library may prove useful for many developers, but I suggest removing the language "with security" and also consider renaming this project.

Thanks for reading!

**Comment by twistedpixel** [comment]
I think rather than renaming the entire project, @samayo could just add a note to the readme file to explain that the project aims to be bulletproof but since that is not possible in software, it should not be treated as the silver bullet of file upload libraries. Really, any sane developer would think like this and continue to treat user-uploaded files as high-risk.

Also I’m pretty sure you can’t execute PHP code without the file extension unless you use eval() or your web server is misconfigured. So just make sure you never add .php to an uploaded file.

**Comment by acicali** [comment]
Unfortunately, many PHP applications use require/include in unsafe ways. A malicious user can upload "myphoto.jpg" to your server with PHP code in it, and trick your application into calling require/include on "myphoto.jpg" (by manipulating the URL for instance).

If your PHP application allows those two things to happen (upload images with embedded PHP + require/include those images) then an attacker can completely own your software or entire server in some cases.

**Comment by samayo** [comment]
Thanks for the feedback, maybe I should start considering about the name like you said, or simply add a warning like @twistedpixel  said. 

But I have one request first, since you mentioned it is possible to trick the application into uploading a malicious file, can you share a code snippet using bulletproof to show how this is done? 

I'm not being standoff-ish, I am just interested to see an example

**Comment by acicali** [comment]
I haven't actually used Bulletproof, but the first example in the README under Usage is a suitable example. Simply put, a valid image can contain PHP code.

Attached is an image that contains PHP code. I simply right-clicked the image in Windows and under the details tab, added PHP for the title.
![wood](https://user-images.githubusercontent.com/353487/45241484-ab283400-b2ba-11e8-91da-64f77ebb79bb.jpg)

Here is how this translates to an attack on the PHP application that uses Bulletproof:

1. Bulletproof allows images to be uploaded that contain PHP, so any PHP application that uses Bulletproof for image uploads also allows an attacker to place PHP code on the application server.

2. Once an attacker has uploaded their malicious image to the server, they must then cause the PHP application to require/include the image file to run their embedded PHP. For instance, an application may insecurely require or include a file based on query string parameters.

--

The above attack vector is a very common scenario. While Bulletproof alone does not provide the complete attack vector, it facilitates number #&#xfeff;1. A new developer that is unfamiliar with how to secure against #&#xfeff;2 is also unlikely to understand that Bulletproof allows #&#xfeff;1.

In summary, someone who searches for "secure php image uploads" may find and use Bulletproof because it's "secure" and "bulletproof". These terms have created a false sense of security for that developer. I suggest that it simply be described as a useful image uploading utility, and include a disclaimer that securely handling image uploads with PHP requires a thorough understanding of best practices and attack vectors.

Please don't take this suggestion as an attack, just feedback. Thanks for your hard work.

**Comment by samayo** [comment]
I have heard this before. :)

Basically what you are saying is, if a user has a pre-existing vulnerability in his code, then when he uses bulletproof ... xy will happen. 

The problem here is the user code. Bulletproof does not introduce false sense of security when it is used on top of unsecure application.

I'm closing this, but feel free to share a real exploit pertinent to this library

**Comment by samayo** [comment]
No it isn't. If you are creating a database wrapper library and someone uses it along outdated and vulnerable version of mysql it does not make your library unsecure.

**Comment by produktive** [comment]
https://www.idontplaydarts.com/2012/06/encoding-web-shells-in-png-idat-chunks/
